FEA Top k accuracy metric

gbolmier opened the pull request:
#### Reference Issues/PRs
Closes #10488
Fixes #10144
Fixes #8234

#### What does this implement/fix? Explain your changes.

This implements a top-k accuracy classification metric, for use with predicted class scores in multiclass classification settings. A prediction is considered top-k accurate if the correct class is one of the k classes with the highest predicted scores.

Commit:
added top_k_accuracy_metric

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Update top_k_caccuracy_score and move to _ranking

Handle multiclass case not multilabel

gbolmier committed:
Update and move test_top_k_accuracy_score to test_ranking

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Fix doc

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+        True labels.
+
+    y_score : array-like of shape (n_samples, n_classes)
+        Target scores.
```
```suggestion
        Target scores, as returned by  :term:`decision_function` or :term:`predict_proba`
```

gbolmier replied:
Done.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    y_score : array-like of shape (n_samples, n_classes)
+        Target scores.
+
+    k : int, optional (default=5)
```
```suggestion
    k : int, default=5
```

same for the other parameters please

gbolmier replied:
Done.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+
+    """
+    check_consistent_length(y_true, y_score)
+    y_type = type_of_target(y_true)
```
I think this should be testing `y_score` instead because it's possible to call the function on a subset of samples that would all be the same class

gbolmier replied:
WDYT of this:
```python
y_type = type_of_target(y_true)

if y_type not in {'binary', 'multiclass'}:
    raise ValueError(
        f"Target type must be 'binary' or 'multiclass', got '{y_type}' "
        "instead."
    )
```

NicolasHug replied:
If that passes a test where `y_true` is e.g. [0, 0, 0], then this should be fine!

gbolmier replied:
Just added a test in the docstring

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    y_score = check_array(y_score)
+    classes = _encode(y_true)
+
+    if len(classes) != y_score.shape[1]:
```
Since all classes might not be represented in the samples, this should probably check `y_true.max()` against `y_score.shape[1]` instead.

gbolmier replied:
Done.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    -----
+    If ``k = 1``, the result will be the same as the accuracy_score (though see
+    note below). If ``k`` is the same as the number of classes, this score will
+    be perfect and meaningless.
```
Let's error if k = 1 or if k = n_classes then

gbolmier replied:
Done.

NicolasHug reviewed (commented):
Thanks for the PR @gbolmier , a few comments ;)

gbolmier committed:
Fix unused import

gbolmier committed:
Update top_k_accuracy_score

Raise errors for `k`=1, `k`=`n_classes`, cover binary `y_true` as it can be a subset of the possible classes, fix doc and update tests and doctest to match changes

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    note below). If ``k`` is the same as the number of classes, this score will
+    be perfect and meaningless.
+
+    In cases where two or more labels are assigned equal probabilities, the
```
Done.

gbolmier commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,88 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    y_type = type_of_target(y_true)
+
+    if y_type != 'multiclass':
+        raise ValueError(f"Target type must be 'multiclass' not {y_type}")
```
Done.

gbolmier commented:
Thank you so much for taking the time of the review @NicolasHug!
Let me know if there is more to improve :)

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    y_score : array-like of shape (n_samples, n_classes)
+        Target scores. These can be either probability estimates or
+        non-thresholded decision values as returned by :term:`predict_proba` or
+        :term:`decision_function`.
```
```suggestion
        :term:`decision_function`, respectively.
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
     return np.average(gain, weights=sample_weight)
+
+
+def top_k_accuracy_score(y_true, y_score, k=5, normalize=True):
```
Thinking about the default, should it be `2`, e.g. the minimum value for which the function can be called? It would work in all cases. 5 would fail if n_classes < 5

gbolmier replied:
Agree with that!

chkoar replied:
I am not sure that we want `ValueError` [here](https://github.com/scikit-learn/scikit-learn/pull/16625/files#diff-2e8e8eee8d9982a7796bdad9558a3482R1502). I believe that for `k>=n_classes` we should raise a warning with the same message probably but we probably want to return a score. No?

NicolasHug replied:
I think we should raise an error if k >= n_classes. This will always output an accuracy of 1 which will lure inexperienced users into thinking their estimator is doing great, where in reality they're just misusing a metric.

In general, we try to make misuses hard / impossible

chkoar replied:
> In general, we try to make misuses hard / impossible  

Agree. 

> will lure inexperienced users into thinking their estimator is doing great

If  they have a binary problem and they ask for top_5 accuracy then their estimator will be doing great obviously!

I dunno. IMHO a warning is enough.

rth replied:
>   If they have a binary problem and they ask for top_5 accuracy then their estimator will be doing great obviously!

Indeed. Also +1 to increase the default to at least 3. I think top_5 is good when you have a very large number of classes (e.g. ImageNet), but 30-60 classes which is more common with tabular data, top_3 is already quite useful.

>

NicolasHug replied:
> If they have a binary problem and they ask for top_5 accuracy then their estimator will be doing great obviously!

Obvious to you, maybe not to them. Also note that we error in the binary case too.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    Returns
+    -------
+    score : float
+        The best performance is 1 with `normalize == True` and the number of
```
```suggestion
        The top-k accuracy score. The best performance is 1 with `normalize == True` and the number of
```

gbolmier replied:
:+1:

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    -----
+    In cases where two or more labels are assigned equal predicted scores
+    labels with smallest indices will be chosen first. Then the result might
+    be incorrect if the good label falls at the treshold because of that.
```
```suggestion
    In cases where two or more labels are assigned equal predicted scores,
    the labels with the smallest indices will be chosen first. Then, the result might
    be incorrect if the correct label falls after the threshold because of that.
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    --------
+    accuracy_score
+
+    Notes
+    -----
```
Hmm, thinking about this more, I think we should typically be permissive in case of ties, and still count it has a hit when we can.

Maybe it would be reasonably easy to implement with the `order` parameter of `argsort`?

gbolmier replied:
Would require `y_score` input to be a numpy array specifying at least one field and adding an optional `order` parameter to `top_k_accuracy_score` that then will be pass to `argsort`.

I just hope that `y_score = check_array(y_score, ensure_min_features=3)` doesn't remove fields from the array.

Should I give a try?

NicolasHug replied:
I think we would need to create the structured array ourselves, after `check_array` is called

gbolmier replied:
After investigating we are actually using [`np.argmax`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.argmax.html#numpy-argmax) in classifiers' `predict` method (cf. [in linear_model](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841/sklearn/linear_model/_base.py#L279)).

`np.argmax` documentation states:
> In case of multiple occurrences of the maximum values, the indices corresponding to the first occurrence are returned.

Should we anyway do something about ties here?

NicolasHug replied:
Let's leave it as-is and wait for others opinion then

jnothman replied:
I would ordinarily err on the side of underestimating performance, rather than being "permissive in case of ties". I think tie breaking in other metrics, such as label ranking metrics like coverage_error, similarly prefers to understate performance.

The current approach provides the benefit that it is consistent with accuracy given scikit-learn classifier conventions. I don't mind it. A conservative tie-breaking approach would also be reasonable.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    if y_true.max() >= y_score.shape[1]:
+        raise ValueError(
+            "Number of classes in 'y_true' is greater than the number of "
+            "columns in 'y_score'."
```
let's print them!
```suggestion
            f"Number of classes in 'y_true' ({blip}) is greater than the number of "
            f"columns in 'y_score' ({blop})."
```

gbolmier committed:
Update top_k_accuracy_score doctest

Add test for case when `y_true` = [0]*4

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,94 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    """
+    if k == 1:
+        raise ValueError(
+            "'k'=1 is simply 'metrics.accuracy_score'. Please, use that "
```
```suggestion
            "'k'=1 is equivalent to 'metrics.accuracy_score'. Please, use that "
```

NicolasHug commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -1468,3 +1469,54 @@ def test_partial_roc_auc_score():
             _partial_roc_auc_score(y_true, y_pred, max_fpr))
+
+
+@pytest.mark.parametrize('k, expected_score', [(2, .2), (3, .4), (4, .8)])
```
let's also check the `normalize=Flase` case?

gbolmier replied:
It's checked in the docstring with the `y_true = [0]*4` case. Do you want me to move them into `test_ranking.py`?

gbolmier replied:
Added both to `test_ranking.py`

NicolasHug reviewed (commented):
Thanks @gbolmier a few more.

Let's also add a  small section in the User Guide!

gbolmier committed:
Update top_k_accuracy_score doc

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

gbolmier committed:
Update top_k_accuracy_score doc

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

gbolmier committed:
Update top_k_accuracy_score

Change `k` default to 2. Update docstring and error msg

gbolmier committed:
Merge remote-tracking branch 'origin/top_k_accuracy_metric' into top_k_accuracy_metric

gbolmier committed:
Fix test typo error

gbolmier committed:
Fix line too long

chkoar commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,100 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
     return np.average(gain, weights=sample_weight)
+
+
+def top_k_accuracy_score(y_true, y_score, k=2, normalize=True):
```
I believe that a reasonable default would be `k=5`.

NicolasHug replied:
thoughts on https://github.com/scikit-learn/scikit-learn/pull/16625#discussion_r387332185 ?

chkoar replied:
I missed your comment.

chkoar commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,100 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+            "therefore meaningless."
+        )
+
+    infered_n_classes = y_true.max() + 1
```
Do we make the assumption that the labels are in `0,1...` or am I missing something?

chkoar commented:
Don't we want a default scorer for that metric? I understand that `k` might be blocker for that.

gbolmier committed:
Update top_k_accuracy_score test

Add `y_true = [0]*5` case

gbolmier committed:
Update top_k_accuracy_score doc

gbolmier committed:
Update top_k_accuracy_score test

Add `normalize == False` case

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Update top_k_accuracy_score

Change `k` default to 3 as it represents better the usual case. Make `y_true` value error message more explicit. Check `y_score` number of columns. Fix `k` check with n_classes

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Fix top_k_accuracy_score docstring typo

gbolmier committed:
Update top_k_accuracy_score

Add a sample_weight parameter

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Fix consistent length check

Check also sample_weight with y_true and y_score

gbolmier committed:
Fix docstring typo

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Add a top-k accuracy score section in the user guide

jnothman commented:
> in a multilabel setting we should compute subset top-k accuracy i.e. each correct output label must be among the top k predicted labels for the corresponding output.

I think subset accuracy at k and recall at k would both be relevant.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,111 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+
+    This metric computes the number of times where the correct label is among
+    the top `k` labels predicted (ranked by predicted scores). Note that
+    multilabel classification case isn't covered here.
```
We typically add references like so:


```suggestion
    multilabel classification case isn't covered here.
    
    Read more in the :ref:`User Guide <top_k_accuracy_score>`
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,111 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    0.75
+    >>> top_k_accuracy_score(y_true, y_score, k=2, normalize=False)
+    3
+    >>> top_k_accuracy_score([0]*4, y_score, k=2)
```
```suggestion
    >>> top_k_accuracy_score([0, 0, 0, 0], y_score, k=2)
```

I'd also make this one the first example this it's the simplest.

Same for the UG

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,111 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics import top_k_accuracy_score
+    >>> y_true = np.array([0, 1, 2, 2])
```
Let's pass it directly as done below instead of declaring.

gbolmier replied:
I'm also using it for the second test, `[0, 0, 0, 0]` is used once in the third test.

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,111 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    if k == 1:
+        raise ValueError(
+            "'k'=1 is equivalent to 'metrics.accuracy_score'. Please, use "
+            "that method instead."
```
Nit
```suggestion
            "that function instead."
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,111 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    if k >= n_classes:
+        raise ValueError(
+            f"'k' ({k}) can't be greater than or equal to 'n_classes' "
+            f"({n_classes}). Will result in a perfect score and is therefore "
```
```suggestion
            f"({n_classes}). This would result in a perfect score and is therefore "
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,111 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    if infered_n_classes > n_classes:
+        raise ValueError(
+            f"Number of classes in 'y_true' {infered_n_classes} is greater "
+            f"than the number of columns in 'y_score' {n_classes}."
```
```suggestion
            f"Number of classes in 'y_true' ({infered_n_classes}) is greater "
            f"than the number of columns in 'y_score' ({n_classes})."
```

NicolasHug commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,46 @@ In the multilabel case with binary label indicators: ::
+`accuracy <https://en.wikipedia.org/wiki/Accuracy_and_precision>`_ in the
+difference that for each sample, ``k`` attempts are allowed to find the correct
+response. Either the fraction (default) or the count (normalize=False) of
+correct predictions is returned.
```
```suggestion
The :func:`top_k_accuracy_score` function is a generalization of :func:`accuracy_score`.
The difference is that a prediction is considered correct as long as the true label is
 associated with one of the ``k`` highest predicted scores. :func:`accuracy_score` is 
 the special case of `k = 1`.
```

NicolasHug commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,46 @@ In the multilabel case with binary label indicators: ::
+
+  >>> import numpy as np
+  >>> from sklearn.metrics import top_k_accuracy_score
+  >>> y_true = np.array([0, 1, 2, 2])
```
same comments as in the docstring but looks good

NicolasHug commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -1468,3 +1469,62 @@ def test_partial_roc_auc_score():
+                                sample_weight) == pytest.approx(true_score)
+
+
+@pytest.mark.parametrize('y_true, k, msg', [
```
We should have a test for the binary case too. 

Which means I doubt we can use the current workaround.

NicolasHug commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -1468,3 +1469,62 @@ def test_partial_roc_auc_score():
+        [0, 1, 2],
+        3,
+        r"'k' \(3\) can't be greater than or equal to 'n_classes' \(3\). Will "
+        "result in a perfect score and is therefore meaningless."
```
Just FYI, you don't need to catch the whole message. The relevant part is enough

```suggestion
        r"'k' \(3\) can't be greater than or equal to 'n_classes' \(3\). Will "
        "result in a perfect score and is therefore meaningless."
```
```suggestion
        r"'k' \(3\) can't be greater than or equal to 'n_classes' \(3\)"
```

gbolmier replied:
Nice! Didn't know it was possible, thx!

NicolasHug commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -1468,3 +1469,62 @@ def test_partial_roc_auc_score():
             _partial_roc_auc_score(y_true, y_pred, max_fpr))
+
+
+@pytest.mark.parametrize('y_true, k, normalize, sample_weight, true_score', [
```
Let's simplify this a bit: 4 classes should be enough, maybe 3?

also maybe group the test by values of k?

NicolasHug reviewed (commented):
Thanks @gbolmier , another round!

gbolmier committed:
Update user guide

gbolmier committed:
Update top_k_accuracy_score

Test y_score number of columns by ourself instead of testing it in check_array and raise a more explicit message error if less than 3. Fix typo. Add a ref of the user guide in the docstring. Simplify tests.

amueller commented on examples/linear_model/plot_poisson_regression_non_normal_loss.py:
```diff
@@ -393,11 +393,11 @@ def _mean_frequency_by_risk_group(y_true, y_pred, sample_weight=None,
 #
 # To compare the 3 models within this perspective, one can plot the fraction of
 # the number of claims vs the fraction of exposure for test samples ordered by
-# the model predictions, from safest to riskiest  according to each model:
```
is this change related?

gbolmier replied:
Nope, got an issue with my last merge from master, just pushed the fix :)

gbolmier committed:
Fix merge error with plot_poisson_regression_non_normal_loss.py example

gbolmier committed:
Update to cover the multilabel case

gbolmier committed:
Update to cover the multilabel case

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier commented:
Sorry I got confused with the [multilabel](https://scikit-learn.org/stable/modules/multiclass.html#multilabel-classification-format) and the [multioutput](https://scikit-learn.org/stable/modules/multiclass.html#multilabel-classification-format) cases, that's why I have only implemented the multiclass setting so far. Now that I have a better view of the problem I can cover both.

I'm concerned about the case where no labels belong to a sample. I think that the prediction scores for this sample shouldn't assign any labels in order to be correct, if we have probabilities we can check that by tresholding at `0.5` but what should we do for non-thresholded decision values as returned by `decision_function`, should we add a threshold parameter?

jnothman commented:
Perhaps for now raise an error if a sample has no labels in the multilabel
case, as this metric is not applicable..?

NicolasHug commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -1468,3 +1469,129 @@ def test_partial_roc_auc_score():
+        [0.3, 0.2, 0.3, 0.2],
+        [0.3, 0.4, 0.1, 0.2],
+        [0.3, 0.4, 0.2, 0.1],
+    ])
```
Is this a relevant output for a multilabel case?

Typically, A DecisionTreeClassifier will output a list of len n_labels where each entry is an array of shape (n_samples, 2)

gbolmier replied:
It depends of the classifier you are using, some output a list of `n_classes` arrays of shape `(n_samples, 2)` as you said and some others output an array of shape `(n_samples, n_classes)`. See @jnothman [gist](https://gist.github.com/jnothman/4807b1b0266613c20ba4d1f88d0f8cf5) for an exhaustive summary.

Since `accuracy_score` expect an array of shape (n_samples, n_classes) for the multilabel case I went for the same solution

NicolasHug replied:
> Since accuracy_score expect an array of shape (n_samples, n_classes) for the multilabel case I went for the same solution

It's quite different since `accuracy_score` expects predicted labels (as returned by `predict`), not predicted scores.

> See @jnothman gist for an exhaustive summary

Thanks, I wasn't even aware of these discrepancies. If we're going to support multilabel then, we should support and test all cases.

Your call, but it might be easier to first merge the PR without multilabel support, and then submit another one.

NicolasHug reviewed (commented):
Thanks @gbolmier , a few comments. I'm not sure about the multilabel case yet

gbolmier committed:
Split multiclass and multilabel tests and code refactoring

gbolmier committed:
Revert "Update to cover the multilabel case"

This reverts commit 0d7da6587a30c7e34d8c8b86c2d856d096ff2392.

gbolmier committed:
Fix revert conflicts to come back to only cover multiclass setting

gbolmier committed:
Fix error doc and tests

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier commented:
As recommended by @NicolasHug let's keep this PR simple by only implementing the multiclass setting. I opened #16666 to discuss the multilabel and multioutput cases

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+
+    This metric computes the number of times where the correct label is among
+    the top `k` labels predicted (ranked by predicted scores). Note that the
+    multilabel and multioutput cases aren't covered here.
```
```suggestion
    multilabel case isn't covered here.
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    -----
+    In cases where two or more labels are assigned equal predicted scores,
+    the labels with the smallest indices will be chosen first. Then, the result
+    might be incorrect if the correct label falls after the threshold because
```
```suggestion
    the labels with the smallest indices will be chosen first. This might impact the result if the correct label falls after the threshold because
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    ----------
+    y_true : array-like of shape (n_samples,)
+        True labels. Expected to be a sequence of int in the range
+        (0, n_classes - 1).
```
```suggestion
        [0, n_classes - 1].
```

`(a, b)` is actually what some of us would write `]a, b[`.

Otherwise we can use "a sequence of ints in `range(n_classes)`"

NicolasHug commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,45 @@ In the multilabel case with binary label indicators: ::
+
+.. math::
+
+   \texttt{top-k accuracy}(y, \hat{y}) = \frac{1}{n_\text{samples}} \sum_{i=0}^{n_\text{samples}-1} \sum_{j=1}^{k} 1(\hat{y}_{i,j} = y_i)
```
Strictly speaking j is in [0, k-1] but I guess it would mostly make the formula a bit ugly so we can keep it as is

NicolasHug commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,45 @@ In the multilabel case with binary label indicators: ::
+  >>> y_score = np.array([[0.5, 0.2, 0.2],
+  ...                     [0.3, 0.4, 0.2],
+  ...                     [0.2, 0.4, 0.3],
+  ...                     [0.7, 0.2, 0.1]])
```
Same for example in docstring

```suggestion
  >>> y_score = np.array([[0.5, 0.2, 0.2],  # 0 is in top 2
  ...                     [0.3, 0.4, 0.2],  # 1 is in top 2
  ...                     [0.2, 0.4, 0.3],  # 2 is in top 2
  ...                     [0.7, 0.2, 0.1]]) # 2 isn't in top 2
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    if n_classes < 3:
+        raise ValueError(
+            "In a multiclass setting the number of columns in 'y_score' must "
+            f"be greater than 2, got ({n_classes}) instead. Please see "
```
Since we're always in a multiclass setting:

```suggestion
            "The number of columns in 'y_score' must "
            f"be strictly greater than 2, got ({n_classes}) instead. Please use "
```

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+            "therefore meaningless."
+        )
+
+    infered_n_classes = y_true.max() + 1
```
They're both inferred so let's call this `n_classes_y_true`

NicolasHug reviewed (approved):
Thanks @gbolmier , looks good!
Minor nits from me.

Let's also add a functional test along the lines of:

```py
def test_top_k_accuracy_increasing():
    # Make sure increasing k leads to a higher score
    X, y = make_classification(n_classes=10, n_samples=1000, n_informative=10, random_state=0)
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

    est = LogisticRegression(random_state=0)
    est.fit(X_train, y_train)

    for X, y in zip((X_train, X_test), (y_train, y_test)):
        scores = [top_k_accuracy_score(y, est.predict_proba(X), k=k) for k in range(2, 10)]
        assert np.all(np.diff(scores) > 0)
```

gbolmier committed:
Fix docstring and add new test

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

NicolasHug commented:
Should be a quick review @adrinjalali @thomasjpfan ?

adrinjalali commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+            "the binary case."
+        )
+
+    if k >= n_classes:
+        raise ValueError(
+            f"'k' ({k}) can't be greater than or equal to 'n_classes' "
+            f"({n_classes}). This will result in a perfect score and is "
+            "therefore meaningless."
+        )
```
I'd consider `k==n_classes` not to be an error. Especially when I was reading the example, I'd be passing k=3 to see if it returns 1 indeed (no strong opinion here though)

NicolasHug replied:
also discussed in https://github.com/scikit-learn/scikit-learn/pull/16625#discussion_r387661189

adrinjalali commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+
+    if y_true_type not in {'binary', 'multiclass'}:
+        raise ValueError(
+            "'y_true' type must be 'binary' or 'multiclass', got "
```
shouldn't accept `binary` if we're not accepting `n_classes=2`.

gbolmier replied:
`y_true` could contains a subset of the classes, e.g. `[0, 0, 1]`

adrinjalali replied:
but then we'll raise because the number of classes in `y_true` is not the same as the number of columns in `y_pred`.

gbolmier replied:
As we assume that `y_true` doesn't have to contain each possible label we are not checking that it has the same number of classes than `y_pred` number of columns. Instead we check that inferred `y_true` number of classes is less or equal to `y_score.shape[1]`

adrinjalali reviewed (commented):
Thanks @gbolmier , here's a fresh review. I haven't read the previous reviews, so my apologies if some of the questions are repeated, happy to be pointed at the right discussion.

NicolasHug commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,45 @@ In the multilabel case with binary label indicators: ::
+correct as long as the true label is associated with one of the ``k`` highest
+predicted scores. :func:`accuracy_score` is the special case of `k = 1`.
+
+The function covers the multiclass classification case but not the binary and
+multilabel cases.
```
This is basically following the zen *There should be one-- and preferably only one --obvious way to do it.*

adrinjalali replied:
to me it's more like `np.average` which calls `np.mean` if sample weights are None.

NicolasHug replied:
that's fair. I don't feel strongly about this one (more so about k == n_classes though). @gbolmier if you decide to change, expect other reviewers to have a different opinion again ;)

gbolmier replied:
I'm a bit torn... I'm concerned about preventing accidental misuses. Maybe raise a warning and call `accuracy_score`

NicolasHug replied:
> I'm concerned about preventing accidental misuses

What kind of misuse with `k=1` do you forsee here?

gbolmier replied:
> > I'm concerned about preventing accidental misuses
> 
> What kind of misuse with `k=1` do you forsee here?

Nevermind, I was thinking of users that could wrongly expect something different than classic accuracy score

gbolmier replied:
> to me it's more like `np.average` which calls `np.mean` if sample weights are None.

Agree, which help code that parametrizes `k` as you pointed in [#16625 (comment)](https://github.com/scikit-learn/scikit-learn/pull/16625#discussion_r404243263)

NicolasHug commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,45 @@ In the multilabel case with binary label indicators: ::
+
+.. math::
+
+   \texttt{top-k accuracy}(y, \hat{y}) = \frac{1}{n_\text{samples}} \sum_{i=0}^{n_\text{samples}-1} \sum_{j=1}^{k} 1(\hat{y}_{i,j} = y_i)
```
I feel like weights complicate the formula for little benefit. Most of the UG doesn't show SWs in the formulas (or show both)

NicolasHug commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
     return np.average(gain, weights=sample_weight)
+
+
+def top_k_accuracy_score(y_true, y_score, k=3, normalize=True,
```
I like 2 better but https://github.com/scikit-learn/scikit-learn/pull/16625#discussion_r387332185

gbolmier committed:
Rewording Top-k Accuracy score section in user guide

gbolmier commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,45 @@ In the multilabel case with binary label indicators: ::
+  ...                     [0.7, 0.2, 0.1]])
+  >>> top_k_accuracy_score(y_true, y_score, k=2)
+  0.75
+  >>> top_k_accuracy_score(y_true, y_score, k=2, normalize=False)
```
Do you mean by adding a comment like this?
```python
# Not normalizing gives the number of correctly classified samples
>>> top_k_accuracy_score(y_true, y_score, k=2, normalize=False) 
```

adrinjalali replied:
yeah exactly, maybe put _correctly_ in quotes since this is a rather loose definition of correctly classified?

gbolmier commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1419,120 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    0.75
+    >>> top_k_accuracy_score(y_true, y_score, k=2, normalize=False)
+    3
+    >>> top_k_accuracy_score([0, 0, 0, 0], y_score, k=2)
```
Agree, let's keep things simple

gbolmier committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into top_k_accuracy_metric

gbolmier committed:
Add comment to top_k_accuracy_score docstring test

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Fix top k accuracy score docstring test bug

thomasjpfan commented:
This PR errors on the binary case and the `k >= n_classes` case. This is the most restrictive ways to handling this. We can release it in its current state and see if users find this behavior unexpected.

Personally, I am coming around to not raising an error. If a user chooses `top_k_accuracy`, I hope they understand the metric enough to know that when `k >= n_classes` the return is one.

adrinjalali commented:
That's exactly why I'd rather have `k <= 2` as the default, and only raise for `k > n_classes`, and support binary as well.

rth commented:
> That's exactly why I'd rather have k <= 2 as the default, and only raise for k > n_classes, and support binary as well.

Let's go with this then. Also checking that it produces asymptotically correct results against other metrics in the binary case or with k=1 is useful.

> If a user chooses top_k_accuracy, I hope they understand the metric enough to know that when k >= n_classes the return is one.

Hopefully writing in as a note in the docstring would be enough.

gbolmier committed:
change top_k_accuracy_score k default to 2

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
add binary case support in top_k_accuracy_score

gbolmier commented:
Nice! Looks like we have reached a consensus, thank you all for your opinions. Then I'm modifying the code to account for @adrinjalali proposed changes

gbolmier committed:
remove  binary check in top_k_accuracy_score

NicolasHug commented:
TBF we usually don't assume users understand what they're doing...

Also as @thomasjpfan suggested, it is safer to restrict first and maybe expand later, that go for the widest support and risk deprecation.

Not gonna argue more though.

rth commented:
> TBF we usually don't assume users understand what they're doing...

I disagree, we should provide safe defaults, explicit documentation and correct computed results. If users do something that doesn't make sense but using valid input & parameters it's on them.
In this particular, case `top_k_accuracy` is defined in the binary case, so it's a bit arbitrary to disallow it.

For k > n_classes, it's a bit more debatable. In a similar situation in PCA we currently raise an error https://github.com/scikit-learn/scikit-learn/blob/9358a6ee8f93511fd615d3264fa7ee9de0f21b93/sklearn/decomposition/_pca.py#L438 . Here though it would mean that top_k_accuracy on some dataset works, but this same metric on a subset of this dataset may fail which is not a great design IMO.   

> it is safer to restrict first and maybe expand later, that go for the widest support and risk deprecation.

I would be OK either way, I just think we should  we have it in the next release.

NicolasHug commented:
> If users do something that doesn't make sense but using valid input & parameters it's on them.


Fair. Though I feel like it's a fine line. For example in slide 11 of [Joel's presentation](https://nlposs.github.io/2018/slides/nothman.pdf), we choose to raise an error because we don't assume all users know what they're doing, i.e. we don't assume they are aware that different average methods exist.

(Also just to be clear and safe, I am **not** saying that "we usually assume users don't know what they're doing", which would be disrespectful. I'm saying we usually try to err on the safe side)

rth commented:
> Fair. Though I feel like it's a fine line

Agreed, there are arguments both for an against it. So what are next steps to get this merged soon ? :)

NicolasHug commented:
the 3 of you are ok with not raising so let's go with that

rth commented:
> the 3 of you are ok with not raising so let's go with that

Maybe raise a warning but not fail if `k >= n_classes`, that's still not a fully normal use case?

adrinjalali commented:
I'd raise a warning/error if `k > n_classes`. `k == n_classes` is still a valid case (although trivial), isn't it?

gbolmier commented:
Since they both output a perfect score of 1, I would say that `k == n_classes`  and `k > n_classes` cases are the same regarding a warning/error policy.

From all the opinions I suggest to raise a warning if `k >= n_classes`, does it sound alright to you?

jnothman commented on doc/modules/model_evaluation.rst:
```diff
@@ -442,6 +449,45 @@ In the multilabel case with binary label indicators: ::
+The function covers the binary and multiclass classification cases but not the
+multilabel case.
+
+If :math:`\hat{y}_{i,j}` is the predicted class corresponding to :math:`j`-th
```
We should rather adopt the notation used below for multilabel ranking metrics, I think. It's very confusing to have `\hat{y}` here indicate a matrix.

jnothman commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1420,114 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+
+    """
+    if k == 1:
+        return accuracy_score(y_true, y_score, normalize, sample_weight)
```
I'm surprised that this could work. accuracy_score doesn't usually take a score matrix...

gbolmier replied:
Yes it doesn't. I added the following in the docstring ```If `k = 1`, :func:`accuracy_score` is called instead.``` and thus assume users to respect `accuracy_score` expected inputs, if not `accuracy_score` will raise

gbolmier replied:
I'm finally modifying for not calling `accuracy_score` as we require scores as input

jnothman commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1420,114 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+            f"than the number of columns in 'y_score' ({n_classes})."
+        )
+
+    sorted_pred = np.argsort(-y_score, axis=1)
```
The tie breaking strategy you propose is not guaranteed by argsort unless the argsort is stable (`kind='mergesort'`). We should probably have more tests to avoid regression on this point.

jnothman replied:
Also, argpartition would suffice algorithmically (but would be slower unless there are very many classes). And reversing the results is cheaper than negating y_score:
```suggestion
    sorted_pred = np.argsort(y_score, axis=1, kind='mergesort')[:, ::-1]
```

gbolmier replied:
If we reverse the results that means that in case of tie, we are now choosing the highest index (instead of the lowest). It's fine for me, let me know if that matters to you

gbolmier replied:
I'm going to separate tie tests from the regular case

jnothman commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1420,114 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+    Parameters
+    ----------
+    y_true : array-like of shape (n_samples,)
+        True labels. Expected to be a sequence of int in the range
```
Ordinarily we would accept string labels. Why not here?

gbolmier replied:
Because we are working with scores. If we accept string labels, then we need to infer which label correspond to each column of `y_score`

gbolmier replied:
My bad, I didn't noticed we were supporting that. I'm going to do as in `roc_auc_score` and add a `labels` parameter

gbolmier replied:
As `type_of_target([1, 2])` returns `'binary'`, do we assume that for all sklearn metrics `y_true` must contains each label at least once?

jnothman reviewed (changes requested):
Thanks for your work @gbolmier. Unfortunately, there are still a few things to resolve.

Apart from anything else, this metric must be added to the common metric tests in `sklearn/metrics/tests/test_common.py`

jnothman commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1420,114 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+        )
+
+    sorted_pred = np.argsort(-y_score, axis=1)
+    hits = [y in pred[:k] for y, pred in zip(y_true, sorted_pred)]
```
Ordinarily, we would use numpy operations to do this. I.e. `(y_true == pred[:, :k]).any(axis=1)`

gbolmier replied:
Doesn't work unfortunately. Couldn't find any numpy operation doing the job. `np.isin` is the closest I found, but it only allow the same set to compare each elements with... any idea?

averdones replied:
I remember seeing something like this somewhere, not sure where: `hits = (y_true == sorted_pred[:, :k].T).any(axis=0)`. Would this work?

gbolmier replied:
Works great! Thanks @averdones :+1:

averdones replied:
Cheers!

jnothman commented on sklearn/metrics/_ranking.py:
```diff
@@ -1419,3 +1420,114 @@ def ndcg_score(y_true, y_score, k=None, sample_weight=None, ignore_ties=False):
+        )
+
+    sorted_pred = np.argsort(-y_score, axis=1)
+    hits = [y in pred[:k] for y, pred in zip(y_true, sorted_pred)]
```
Ordinarily, we would use numpy operations to do this. I.e. `(y_true == pred[:, :k]).any(axis=1)`

gbolmier replied:
Duplicate with [here](https://github.com/scikit-learn/scikit-learn/pull/16625#discussion_r410911419).

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Notation change for top_k_accuracy_score in the user guide

Change y hat to f hat for consistency with multilabel ranking metrics

gbolmier committed:
Update top_k_accuracy_score

Add a labels parameter. Handle binary case instead of calling `accuracy_score`. Add support for string labels. Guarantee the tie breaking strategy — which choose now the highest indices labels first. Update tests according to the changes.

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Fix PEP8 violations

gbolmier commented:
Thanks a lot for taking the time of the review @jnothman!

> Apart from anything else, this metric must be added to the common metric tests in `sklearn/metrics/tests/test_common.py`

I encountered some issues here. The metric should belong to `THRESHOLDED_METRICS`, but a bunch of tests requires to support `y_score` with shape (n_samples,) — which is not possible as we need all scores to order them. Am I missing something?

Also, when I generate the doc, `sklearn.metrics.top_k_accuracy_score` page is blank. I guess I have to add the metric somewhere in a doc config file?

gbolmier committed:
Remove comment making test fails

gbolmier committed:
Update top_k_accuracy_score

Use numpy operations to compute the hits

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Update top_k_accuracy_score

Add kw argument only. Fix specific tests that were failing because of kw argument only. Add the metric in common tests.

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Update top_k_accuracy_score

Fix _encode and _unique api changes

gbolmier committed:
Add top_k_accuracy_score entry to the 0.24 changelog

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier commented:
> > Apart from anything else, this metric must be added to the common metric tests in `sklearn/metrics/tests/test_common.py`
> 
> I encountered some issues here. The metric should belong to `THRESHOLDED_METRICS`, but a bunch of tests requires to support `y_score` with shape (n_samples,)

I did add the metric to `THRESHOLDED_METRICS`. To avoid failures, in many tests (see changes in `test_common.py`) that expect `y_score` can be 1D, I created a `METRICS_REQUIRE_2D_Y_SCORE`, and added if statements in the tests concerned.

I'm not a huge fan because it adds an additional layer of complexity to the tests, but couldn't think of a better solution, any suggestions?

BTW, I don't understand how `y_score` in other metrics could be 1D, even in the binary case we need both scores to discriminate them as we are not only dealing with probabilities. What am I missing?

> Also, when I generate the doc, `sklearn.metrics.top_k_accuracy_score` page is blank. I guess I have to add the metric somewhere in a doc config file?

I do have an entry in `doc/modules/classes.rst`, but when I generate the doc, the link isn't working:

![Screenshot from 2020-05-14 10-03-17](https://user-images.githubusercontent.com/25319692/81944155-575f0980-95ca-11ea-9ca8-84312f7c447b.png)

Thanks for your help :)

amueller commented:
Looks like it renders correctly on CI: https://103206-843222-gh.circle-artifacts.com/0/doc/modules/classes.html

amueller commented:
Re 1d: this metric, similar to others only requires the ranking of the data. For binary, the ``decision_function`` returns a 1d array.

jnothman commented:
>  I created a METRICS_REQUIRE_2D_Y_SCORE

This is already implied by the combination of `THRESHOLDED_METRICS` and `METRIC_UNDEFINED_BINARY`

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Update top_k_accuracy_score to expect 1D y_score in the binary case

gbolmier committed:
Fix top_k_accuracy_score example comment

amueller commented:
please fix the merge issue

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

pengxin99 commented:
Hi, @gbolmier looks pretty work. Just wonder that will this top-k metric provide global update? as follows:

```
for _input, _label in zip(test_dataset):
    _predict = predict_function(_input)
    top_k_acc = top_k_accuracy_score(_predict, _label)

print(top_k_acc)
```

gbolmier commented:
Hi @pengxin99! Nope, you need to give to the metric all the data at once (same for all other metrics in sklearn)

pengxin99 commented:
> Hi @pengxin99! Nope, you need to give to the metric all the data at once (same for all other metrics in sklearn)

Ok, thanks :)

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Test top_k_accuracy_score normalize option in test_common.py

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Fix user guide

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier commented:
Should be better like that @jnothman

rth commented on sklearn/metrics/_ranking.py:
```diff
@@ -1441,3 +1441,151 @@ def ndcg_score(y_true, y_score, *, k=None, sample_weight=None,
+            y_pred = (y_score > threshold).astype(np.int)
+            hits = y_pred == y_true_encoded
+        else:
+            hits = [True] * len(y_score)
```
```suggestion
            hits =np.one_like(y_score, dtype=np.bool_)
```
to avoid unnecessary python lists

rth commented on sklearn/metrics/_ranking.py:
```diff
@@ -1441,3 +1441,151 @@ def ndcg_score(y_true, y_score, *, k=None, sample_weight=None,
+        labels in ``y_true``.
+
+    k : int, default=2
+        Number of guesses allowed to find the correct label.
```
I'm not sure the the word "guess" is appropriate here. It implies sampling, while this approach is fully deterministic. Maybe

```suggestion
        Number of most likely outcomes considered to find the
        correct label.
```

rth commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -240,7 +241,10 @@ def precision_recall_curve_padded_thresholds(*args, **kwargs):
+    "dcg_score": dcg_score,
+
+    "top_k_accuracy_score": top_k_accuracy_score,
+    "top_k_accuracy_score_binary": partial(top_k_accuracy_score, k=1)
```
Could someone please comment why this is necessary. I couldn't find a reference to it in the above discussion. Wouldn't it be equal to `accuracy_score` except that it takes a probability instead of the class label as input? What's the use case for it?

Also it doesn't follow the convention "*_score"

gbolmier replied:
Thanks for raising this point Roman! The reason was because `test_binary_sample_weight_invariance` naturally fails for any `k>1`. My fix was clumsy as it doubled some tests and needed to add `top_k_accuracy_score` to `METRIC_UNDEFINED_BINARY`.

Instead, I removed `top_k_accuracy_score_binary` and added the following to `test_binary_sample_weight_invariance`:

```python
# top_k_accuracy_score always lead to a perfect score for k > 1 in the
# binary case
metric = partial(metric, k=1) if name == "top_k_accuracy_score" else metric
```

rth reviewed (commented):
Thanks for your work and patience @gbolmier, a few more minor comments below otherwise looks good.

Also please add it to https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values

gbolmier committed:
Rephrase docstring and avoid unnecessary python lists

# More detailed explanatory text, if necessary. Wrap it to about 72
# characters or so. In some contexts, the first line is treated as the
# subject of the commit and the rest of the text as the body. The
# blank line separating the summary from the body is critical (unless
# you omit the body entirely); various tools like `log`, `shortlog`
# and `rebase` can get confused if you run the two together.

# Explain the problem that this commit is solving. Focus on why you
# are making this change as opposed to how (the code explains that).
# Are there side effects or other unintuitive consequences of this
# change? Here's the place to explain them.

# Further paragraphs come after blank lines.

#  - Bullet points are okay, too

#  - Typically a hyphen or asterisk is used for the bullet, preceded
#    by a single space, with blank lines in between, but conventions
#    vary here

# If you use an issue tracker, put references to them at the bottom,
# like this:

# Resolves: #123
# See also: #456, #789

gbolmier committed:
Fix common tests

# More detailed explanatory text, if necessary. Wrap it to about 72
# characters or so. In some contexts, the first line is treated as the
# subject of the commit and the rest of the text as the body. The
# blank line separating the summary from the body is critical (unless
# you omit the body entirely); various tools like `log`, `shortlog`
# and `rebase` can get confused if you run the two together.

# Explain the problem that this commit is solving. Focus on why you
# are making this change as opposed to how (the code explains that).
# Are there side effects or other unintuitive consequences of this
# change? Here's the place to explain them.

# Further paragraphs come after blank lines.

#  - Bullet points are okay, too

#  - Typically a hyphen or asterisk is used for the bullet, preceded
#    by a single space, with blank lines in between, but conventions
#    vary here

# If you use an issue tracker, put references to them at the bottom,
# like this:

# Resolves: #123
# See also: #456, #789

gbolmier committed:
Add a scorer for model evaluation

# More detailed explanatory text, if necessary. Wrap it to about 72
# characters or so. In some contexts, the first line is treated as the
# subject of the commit and the rest of the text as the body. The
# blank line separating the summary from the body is critical (unless
# you omit the body entirely); various tools like `log`, `shortlog`
# and `rebase` can get confused if you run the two together.

# Explain the problem that this commit is solving. Focus on why you
# are making this change as opposed to how (the code explains that).
# Are there side effects or other unintuitive consequences of this
# change? Here's the place to explain them.

# Further paragraphs come after blank lines.

#  - Bullet points are okay, too

#  - Typically a hyphen or asterisk is used for the bullet, preceded
#    by a single space, with blank lines in between, but conventions
#    vary here

# If you use an issue tracker, put references to them at the bottom,
# like this:

# Resolves: #123
# See also: #456, #789

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into top_k_accuracy_metric

gbolmier commented:
> Thanks for your work and patience @gbolmier, a few more minor comments below otherwise looks good.

This is how you learn the ropes :), thanks a lot for the review @rth! Let me know if it needs any other changes

rth committed:
Merge branch 'master' into top_k_accuracy_metric

rth reviewed (approved):
Thanks @gbolmier! LGTM. Should be good to merge unless anyone has more comments?

@jnothman I think you review comments were addressed.

cmarmo commented:
Hi @jnothman, do you mind checking if your comments have been addressed? Two approvals here already. Thanks for your time.

cmarmo commented:
Two approvals here! Time to merge?

rth committed:
Merge branch 'master' into top_k_accuracy_metric

rth commented:
Merging thanks again @gbolmier ! Also thank you for the follow up @cmarmo !

jnothman commented:
Thanks @gblomier. Sorry I've not found time to follow up!
